python/lsst/pipe/base/butlerQuantumContext.py

```diff
@@ -0,0 +1,204 @@
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Module defining a butler like object specified to a specific quantum.
+"""
+
+__all__ = ("ButlerQuantumContext",)
+
+from enum import Enum, auto
+import types
+import typing
+
+from .connections import InputQuantizedConnection, OutputQuantizedConnection, DeferredDatasetRef
+from .struct import Struct
+from lsst.daf.butler import DatasetRef, Butler, Quantum
+
+
+class RefDirection(Enum):
+    IN = auto()
+    OUT = auto()
+
+
+class ButlerQuantumContext:
+    """Butler like class specialized for a single quantum
+
+    A ButlerQuantumContext class wraps a standard butler interface and
+    specializes it to the context of a given quantum. What this means
+    in practice is that the only gets and puts that this class allows
+    are DatasetRefs that are contained in the quantum.
+
+    In the future this class will also be used to record providence on
+    what was actually get and put. This is in contrast to what the
+    preflight expects to be get and put by looking at the graph before
+    execution.
+
+    Parameters
+    ----------
+    butler : `lsst.daf.butler.Butler`
+        Butler object from/to which datasets will be get/put
+    quantum : `lsst.daf.butler.core.Quantum`
+        Quantum object that describes the data-sets which will
+        be get/put by a single execution of this node in the
+        pipeline graph.
+    """
+    def __init__(self, butler: Butler, quantum: Quantum):
+        self.quantum = quantum
+        self.registry = butler.registry
+        self.allInputs = set()
+        self.allOutputs = set()
+        for refs in quantum.predictedInputs.values():
+            for ref in refs:
+                self.allInputs.add((ref.datasetType, ref.dataId))
+        for refs in quantum.outputs.values():
+            for ref in refs:
+                self.allOutputs.add((ref.datasetType, ref.dataId))
+
+        # Create closures over butler to discourage anyone from directly
+        # using a butler reference
+        def _get(self, ref):
+            if isinstance(ref, DeferredDatasetRef):
+                self._checkMembership(ref.datasetRef, RefDirection.IN)
+                return butler.getDeferred(ref.datasetRef)
+
+            else:
+                self._checkMembership(ref, RefDirection.IN)
+                return butler.get(ref)
+
+        def _put(self, value, ref):
+            self._checkMembership(ref, RefDirection.OUT)
+            butler.put(value, ref)
+
+        self._get = types.MethodType(_get, self)
+        self._put = types.MethodType(_put, self)
+
+    def get(self, dataset: typing.Union[InputQuantizedConnection, typing.List[DatasetRef], DatasetRef]):
+        """Fetches data from the butler
+
+        Parameters
+        ----------
+        dataset : `InputQuantizedConnection` or `list` of `lsst.daf.butler.DatasetRef`
```

Comment:
I think our latest convention is to write it as 

    `list` [`lsst.daf.butler.DatasetRef`]